Objects/abstract.c

```diff
@@ -2881,7 +2881,50 @@ PyAIter_Check(PyObject *obj)
             tp->tp_as_async->am_anext != &_PyObject_NextNotImplemented);
 }
 
+static int
+iternext(PyObject *iter, PyObject **item)
+{
+    iternextfunc tp_iternext = Py_TYPE(iter)->tp_iternext;
+    if ((*item = tp_iternext(iter))) {
+        return 1;
+    }
+
+    PyThreadState *tstate = _PyThreadState_GET();
+    /* When the iterator is exhausted it must return NULL;
+     * a StopIteration exception may or may not be set. */
+    if (!_PyErr_Occurred(tstate)) {
+        return 0;
+    }
+    if (_PyErr_ExceptionMatches(tstate, PyExc_StopIteration)) {
+        _PyErr_Clear(tstate);
+        return 0;
+    }
+
+    /* Error case: an exception (different than StopIteration) is set. */
+    return -1;
+}
+
+/* Return 1 and set 'item' to the next item of 'iter' on success.
+ * Return 0 and set 'item' to NULL when there are no remaining values.
+ * Return -1, set 'item' to NULL and set an exception on error.
+ */
+int
+PyIter_NextItem(PyObject *iter, PyObject **item)
+{
+    assert(iter != NULL);
+    assert(item != NULL);
+
+    if (!PyIter_Check(iter)) {
```

Comment:
This is more expensive than it needs to be. The simpler `Py_TYPE(iter)->tp_iternext != NULL` is enough as `_PyObject_NextNotImplemented` will raise an exception that needs handling anyway.

To avoid fetching `tp_iternext ` twice, maybe move this check into iternext.

Reply:
True, checking `tp_iternext` should be enough.

> To avoid fetching `tp_iternext ` twice, maybe move this check into iternext.

That will slightly change the behaviour of `PyIter_Next` also, as it currently does not perform this check. Maybe that's ok.

Reply:
I reduced the check to a null pointer check.